### ISSUE TITLE
Better typing for Elements?

### ISSUE THREAD

**Issue by mikeplus64** [ryansolid/solid#190]
In particular it would be nice to have an element type that has the props as a generic parameter, like how `React.Element` does. This allow for imo much safer/more accurate types for e.g. `Switch`. At the moment every component just awkwardly returns a `JSX.Element` regardless of what it actually is. Is this is a limitation of non-react jsx support in typescript?

**Comment by ryansolid** [comment]
Pretty much from my understanding. They expect everything to be of a generic node type. This is even more annoying with intrinsic elements where Solid would clearly benefit from:

```jsx
const div = <div/>; // this should be an HTMLDivElement
```

Something might be possible here that I'm not aware of. There are a number of open issues around TSX that seem to refer to parts of this problem, but there are a lot of subtleties. As far as I know React has the same limitation though. Although given their JSX generally returns VDOM nodes it isn't a big deal.

As far as I know `React.Element` isn't a generic. Are you referring to how you can do that with Components?... Solid has a component type when authoring. Like: 

```jsx
import { Component } from "solid-js";

const MyComp: Component<MyCompProps>(props) { .... } 
```
Although I don't think this helps with Children.

Maybe someone a bit more experienced with TS can help clarify this a bit?

**Comment by mikeplus64** [comment]
>As far as I know React.Element isn't a generic.
It is actually; see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/491f394a3ee3432d00284bbc204e775e3dcc5ca5/types/react/index.d.ts#L146

Plain HTML elements (as in e.g. `<div/>`) do also get quite granular types in React. I think this is due to their design of their JSX factory rather than any magical treatment by TypeScript; see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/491f394a3ee3432d00284bbc204e775e3dcc5ca5/types/react/index.d.ts#L2657.

Does Solid do anything special with the Babel plugin to transform jsx? I understand atm the way to compile `.tsx` with Solid is through TS's "preserve" mode which just outputs a `.jsx` (which then babel is happy to consume) -- if there's nothing magic in there then maybe all that needs to be done is to define a React-esque `jsxFactory` and `jsxFragmentFactory`.

**Comment by mikeplus64** [comment]
Some more inspiration can be found with Preact, probably, see https://github.com/preactjs/preact/blob/master/src/index.d.ts and https://github.com/preactjs/preact/blob/master/src/jsx.d.ts =)

**Comment by ryansolid** [comment]
Yeah the types are actually based on Preacts and Infernos as far as I could take it.

Ah.. yes `React.Element` is generic but React's `JSX.Element` isn't. It just says that it is a `React.Element` with any.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/491f394a3ee3432d00284bbc204e775e3dcc5ca5/types/react/index.d.ts#L2940

`React.Element` is a VDOM node as is Preacts `VNode`. They are basically the output from the JSX not the input. So like restricting props is handled by the function definition of the component, and any props on `React.Element` are just there to define the output. But Solid's output is nothing specific. It's a DOM node, or array of DOM Node, or a function, or a string etc.. there isn't a specific type that is returned from a Component.

I'm not sure how that bit with ReactHTML works but their nodes coming back are VDOM nodes so I'm not sure they'd ever be in the Div scenario I was talking about. There are a couple open typescript issues about this. I'd need to dig them up.

Which as you can guess by my answers Babel's plugin to transform JSX is completely different. It looks like JSX but it might as well be Svelte. In order to do accomplish what the library needs I do completely different transforms, as simple HyperScript would be inefficient to handle reactive wrapping or real DOM creation. Standard JSX to HyperScript transformation is designed for repeated top-down re-execution.

**Comment by mikeplus64** [comment]
>Which as you can guess by my answers Babel's plugin to transform JSX is completely different. It looks like JSX but it might as well be Svelte. In order to do accomplish what the library needs I do completely different transforms, as simple HyperScript would be inefficient to handle reactive wrapping or real DOM creation. Standard JSX to HyperScript transformation is designed for repeated top-down re-execution.

Drat. Still, it might be beneficial to have a custom `jsxFactory` that emits code that's immediately gobbled up by the Babel plugin, but for producing nice types for the user in TypeScript-land.

>That being said I do wonder if there is something that can be done for Switch/Match. Everytime I've tried it's complained I wasn't returning a JSX.Element unless I changed the definition to be really wide (basically any)

Personally I just don't bother with `Switch`/`Match` but use the following:

```typescript

export default function Switch<
  K extends string | number | symbol,
  Options extends { [k in K]: JSX.Element },
>(props: {
  fallback?: JSX.Element,
  value: K,
  options: Options
}) {
  return suspend(() => {
    return props.options[props.value] ?? props.fallback;
  });
}

// example
// currentRoute.name: 'home' | 'profile' | 'self';
<Switch 
  value={currentRoute.name} 
  options={{
    home: <Home />,
    profile: <Profile id={currentRoute.params.id} />,
    self: <Self />
  }} 
/>
```

My head is not fully around how suspense/reactivity works in Solid yet so I am not sure if this has any drawbacks over the default.

There might be a way to do it now with `children` with some faking of nominal types, like

```typescript
interface IsMatch {__isMatch: void}
function Match(): JSX.Element & IsMatch { ...; return element as JSX.Element & IsMatch; }
interface SwitchProps { children: (JSX.Element & IsMatch)[] }
function Switch(props: SwitchProps) {...}
```

**Comment by ryansolid** [comment]
Hmm.. I guess the one thing I'm missing, is outside of the `Switch` `Match` case where there are specific children, which you aren't using, what's the missing part? As far as I know typing works reasonably well? Like it properly checks props on Components etc.. From JSX perspective `JSX.Element` is `React.Element<any, any>` in React so while it always produces a VDOM node generally it is a background thing as end users don't interact with them much?

The types are fairly extensive right down to proper attributes for each type of Intrinsic Element(https://github.com/ryansolid/solid/blob/master/packages/solid/src/rendering/jsx.ts). I never really thought much of what the VDOM node was doing. Like what about the React solution works differently that you are looking for in terms of typing?

**Comment by mikeplus64** [comment]
> Hmm.. I guess the one thing I'm missing, is outside of the `Switch` `Match` case where there are specific children, which you aren't using, what's the missing part? As far as I know typing works reasonably well? Like it properly checks props on Components etc.. From JSX perspective `JSX.Element` is `React.Element<any, any>` in React so while it always produces a VDOM node generally it is a background thing as end users don't interact with them much?
> 
> The types are fairly extensive right down to proper attributes for each type of Intrinsic Element(https://github.com/ryansolid/solid/blob/master/packages/solid/src/rendering/jsx.ts). I never really thought much of what the VDOM node was doing. Like what about the React solution works differently that you are looking for in terms of typing?

Ah, I thought that the type inference for `JSX.Element` would basically constrain `React.ReactElement<any, any>`, to whatever the actual type of `children` was but looks like I was mistaken. `any` always wins =D. There's an ongoing TypeScript issue here https://github.com/Microsoft/TypeScript/issues/14729 .

As for bikeshedding `Switch`/`Match IMO just don't use `children` at all, or type `children` as something other than `JSX.Element` which doesn't defeat type inference -- like this is perfectly legal and does prevent invalid "children":

```typescript
function Switch(props: { ..., children: [boolean, JSX.Element][] }) {...}
<Switch>{[
  [foo===bar, <FooBar/>],
  [baz===bar, <BazBar/>],
  [foo===baz, <FooBaz/>],
  [true, 'dunno'],
]}</Switch>
```

(In that case I don't really see the point of using `children` at all as opposed to just an ordinary prop, but there are libraries in the wild doing it e.g. `react-virtualized`)

**Comment by ryansolid** [comment]
Yeah.. unfortunately ergonomics and clear syntax cues are going to win out against TS every time. Solid's compiler look at the JSX structure to determine how to wrap reactive expressions. Moving to all objects would require the end user to wrap things themselves to retain granularity to prevent unnecessary DOM creation as it is much harder(impossible) to analyse intent.

Like in your examples you'd be creating the other non-travelled routes unnecessarily since to resolve the array or object you resolve all properties. You can avoid this by wrapping each path with a function you execute when the condition is met. However using separate props on children components the compiler knows to do that automatically. This automatic propagation of lazy evaluation is key.

Render props are fine though instead of Children but still problem if rendering JSX in them as still just JSX.Element.

Generally I've found JSX a weaker area for TS and sort of caution people that TS while mature in other areas really isn't here. There are some spec features they don't support at all like namespaces, which could make a big difference for library like this. There are limitations on typing. Certain rules around identifying intrinsic elements that are too restrictive to handle natural patterns for alternate renderers like Native. Lack of consensus on Fragment support. Typing tied to JSXFactory unless you are willing pollute global namespace, or force an import in every file. I'm sure TS will get there one day.
